lib/color_spinor_util.cu

```diff
@@ -120,89 +126,94 @@ namespace quda {
-      genericSource<Float,nSpin,20,order>(a,sourceType, x, s, c);
+      genericSource<Float,nSpin,20,order>(pack);
     } else if (a.Ncolor() == 24) {
-      genericSource<Float,nSpin,24,order>(a,sourceType, x, s, c);
+      genericSource<Float,nSpin,24,order>(pack);
 #ifdef NSPIN4
     } else if (a.Ncolor() == 32) {
-      genericSource<Float,nSpin,32,order>(a,sourceType, x, s, c);
+      genericSource<Float,nSpin,32,order>(pack);
 #endif // NSPIN4
 #ifdef NSPIN1
     } else if (a.Ncolor() == 64) {
-      genericSource<Float,nSpin,64,order>(a,sourceType, x, s, c);
+      genericSource<Float,nSpin,64,order>(pack);
     } else if (a.Ncolor() == 96) {
-      genericSource<Float,nSpin,96,order>(a,sourceType, x, s, c);
+      genericSource<Float,nSpin,96,order>(pack);
 #endif // NSPIN1
 #endif // GPU_MULTIGRID
     } else {
-      errorQuda("Unsupported nColor=%d\n", a.Ncolor());
+      errorQuda("Unsupported nColor=%d", a.Ncolor());
     }
   }
 
-  template <typename Float, QudaFieldOrder order>
-  void genericSource(cpuColorSpinorField &a, QudaSourceType sourceType, int x, int s, int c) {
+  template <typename Float, QudaFieldOrder order, typename pack_t>
+  void genericSource(const pack_t &pack)
+  {
+    auto &a = std::get<0>(pack);
     if (a.Nspin() == 1) {
 #ifdef NSPIN1
-      genericSource<Float,1,order>(a,sourceType, x, s, c);
+      genericSource<Float,1,order>(pack);
 #else
       errorQuda("nSpin=1 not enabled for this build");
 #endif
     } else if (a.Nspin() == 2) {
 #ifdef NSPIN2
-      genericSource<Float,2,order>(a,sourceType, x, s, c);
+      genericSource<Float,2,order>(pack);
 #else
       errorQuda("nSpin=2 not enabled for this build");
 #endif
     } else if (a.Nspin() == 4) {
 #ifdef NSPIN4
-      genericSource<Float,4,order>(a,sourceType, x, s, c);
+      genericSource<Float,4,order>(pack);
 #else
       errorQuda("nSpin=4 not enabled for this build");
 #endif
     } else {
-      errorQuda("Unsupported nSpin=%d\n", a.Nspin());
+      errorQuda("Unsupported nSpin=%d", a.Nspin());
     }
   }
 
-  template <typename Float>
-  void genericSource(cpuColorSpinorField &a, QudaSourceType sourceType, int x, int s, int c) {
+  template <typename Float, typename pack_t>
+  void genericSource(const pack_t &pack)
+  {
+    auto &a = std::get<0>(pack);
     if (a.FieldOrder() == QUDA_SPACE_SPIN_COLOR_FIELD_ORDER) {
-      genericSource<Float,QUDA_SPACE_SPIN_COLOR_FIELD_ORDER>(a,sourceType, x, s, c);
+      genericSource<Float,QUDA_SPACE_SPIN_COLOR_FIELD_ORDER>(pack);
     } else {
-      errorQuda("Unsupported field order %d\n", a.FieldOrder());
+      errorQuda("Unsupported field order %d", a.FieldOrder());
     }
-
   }
 
-  void genericSource(cpuColorSpinorField &a, QudaSourceType sourceType, int x, int s, int c) {
-
+  void genericSource(cpuColorSpinorField &a, QudaSourceType sourceType, int x, int s, int c)
+  {
+    using pack_t = std::tuple<cpuColorSpinorField&, QudaSourceType, int, int, int>;
```

Comment:
Why did we switch to creating an `std::tuple` from passing all of the arguments along? Was this just a design experiment?

Reply:
This was indeed a design experiment to clean up some of the excessive parameter passing.  Enacting this experiment more broadly, or some other alternative approach is deferred until post GK.

Reply:
Fair enough, I think we should revert it instead of leaving a design experiment around, but I'm also okay with leaving it as is for now.